File: backend/app/services/workspace_service.py
from uuid import UUID

from app.core.exceptions import ResourceNotFoundError
from app.models.user import User
from app.models.workspace import Workspace
from app.models.workspace_member import (WorkspaceMember,WorkspaceRole,)
from app.repositories.workspace_member_repository import (WorkspaceMemberRepository,)
from app.repositories.workspace_repository import WorkspaceRepository
from app.schemas.workspace import (
    WorkspaceCreate,
    WorkspaceUpdate,
)


class WorkspaceService:
    def __init__(
        self,
        repository: WorkspaceRepository,
        member_repository: WorkspaceMemberRepository,
    ):
        self.repository = repository
        self.member_repository = member_repository

    async def create_workspace(
        self,
        owner: User,
        data: WorkspaceCreate,
    ) -> Workspace:
        workspace = Workspace(
            name=data.name,
            description=data.description,
            owner_id=owner.id,
        )
        
        workspace = await self.repository.create(workspace)
        
        owner_member = WorkspaceMember(
        workspace_id=workspace.id,user_id=owner.id,role=WorkspaceRole.OWNER,)
        
        await self.member_repository.create(owner_member)

        return workspace

    async def list_workspaces(
        self,
        owner: User,
    ) -> list[Workspace]:
        return await self.repository.get_all_by_owner(
            owner.id,
        )

    async def get_workspace(
        self,
        owner: User,
        workspace_id: UUID,
    ) -> Workspace:
        workspace = await self.repository.get_by_id(
            workspace_id,
        )

        if (
            workspace is None
            or workspace.owner_id != owner.id
        ):
            raise ResourceNotFoundError(
                "Workspace not found."
            )

        return workspace

    async def update_workspace(
        self,
        owner: User,
        workspace_id: UUID,
        data: WorkspaceUpdate,
    ) -> Workspace:
        workspace = await self.get_workspace(
            owner,
            workspace_id,
        )

        update_data = data.model_dump(
            exclude_unset=True,
        )

        for key, value in update_data.items():
            setattr(workspace, key, value)

        return await self.repository.update(workspace)

    async def delete_workspace(
        self,
        owner: User,
        workspace_id: UUID,
    ) -> None:
        workspace = await self.get_workspace(
            owner,
            workspace_id,
        )

        await self.repository.delete(workspace)

File: frontend/services/workspace.service.ts
import { apiClient } from "@/lib/api-client";
import { WORKSPACES } from "@/constants";

export interface WorkspaceCreate {
  name: string;
  description?: string;
}

export interface WorkspaceUpdate {
  name?: string;
  description?: string;
}

export interface WorkspaceResponse {
  id: string;
  name: string;
  description: string | null;
  owner_id: string;
  created_at: string;
  updated_at: string;
}

export interface InviteMemberRequest {
  email: string;
}

export interface AcceptInvitationRequest {
  token: string;
}

export type WorkspaceRole = "owner" | "admin" | "member";

export interface WorkspaceMemberResponse {
  id: string;
  workspace_id: string;
  user_id: string;
  role: WorkspaceRole;
  created_at: string;
  updated_at: string;
}

export interface WorkspaceMemberRoleUpdate {
  role: WorkspaceRole;
}

export interface WorkspaceInvitationResponse {
  id: string;
  workspace_id: string;
  email: string;
  token: string;
  expires_at: string;
  accepted: boolean;
  created_at: string;
}

export const workspaceService = {
  createWorkspace: async (data: WorkspaceCreate): Promise<WorkspaceResponse> => {
    const response = await apiClient.post(WORKSPACES.CREATE, data);
    return response.data;
  },

  listWorkspaces: async (): Promise<WorkspaceResponse[]> => {
    const response = await apiClient.get(WORKSPACES.LIST);
    return response.data;
  },

  getWorkspace: async (workspaceId: string): Promise<WorkspaceResponse> => {
    const response = await apiClient.get(WORKSPACES.DETAILS(workspaceId));
    return response.data;
  },

  updateWorkspace: async (workspaceId: string, data: WorkspaceUpdate): Promise<WorkspaceResponse> => {
    const response = await apiClient.patch(WORKSPACES.UPDATE(workspaceId), data);
    return response.data;
  },

  deleteWorkspace: async (workspaceId: string): Promise<void> => {
    const response = await apiClient.delete(WORKSPACES.DELETE(workspaceId));
    return response.data;
  },

  listMembers: async (workspaceId: string): Promise<WorkspaceMemberResponse[]> => {
    const response = await apiClient.get(WORKSPACES.MEMBERS(workspaceId));
    return response.data;
  },

  changeMemberRole: async (workspaceId: string, userId: string, data: WorkspaceMemberRoleUpdate): Promise<WorkspaceMemberResponse> => {
    const response = await apiClient.patch(WORKSPACES.MEMBER_ROLE(workspaceId, userId), data);
    return response.data;
  },

  removeMember: async (workspaceId: string, userId: string): Promise<{ message: string }> => {
    const response = await apiClient.delete(WORKSPACES.REMOVE_MEMBER(workspaceId, userId));
    return response.data;
  },

  inviteMember: async (workspaceId: string, data: InviteMemberRequest): Promise<WorkspaceInvitationResponse> => {
    const response = await apiClient.post(WORKSPACES.INVITE(workspaceId), data);
    return response.data;
  },

  acceptInvitation: async (data: AcceptInvitationRequest): Promise<WorkspaceMemberResponse> => {
    const response = await apiClient.post(WORKSPACES.ACCEPT_INVITATION, data);
    return response.data;
  },
};
